### PR TITLE
Unify PySCF and GPU4PySCF nuclear-gradient computation

### PR DESCRIPTION
**Pull request by JensWehner**
- Add a shared feature_derivatives() implementation based on torch.autograd.grad().
- Add grid_derivative_block() to centralize slicing of grid-resolved derivatives.
  - Handles point-first layouts for grid coordinates and weights.
  - Handles point-last layouts for density, density gradient, and kinetic density.
- Add contract_veff_block() for the shared AO/XC derivative contractions.
  - Documents PySCF’s packed AO derivative layout.
  - Shares density, gradient, and kinetic-energy contractions between CPU and GPU paths.
- Add focused tests for feature differentiation, grid-block slicing, and effective-potential contraction against hardcoded reference values.
- Update the supported Pixi version from 0.75 to 0.78 and refresh the lockfile.

### CODE REVIEW

File: .github/workflows/skala-test.yml [hunk 1]
```diff
@@ -50,7 +50,7 @@ jobs:
     - name: Setup Pixi
       uses: prefix-dev/setup-pixi@f00437f565399d418b0acc85936d12c1fb668347 # v0.10.1
       with:
-        pixi-version: v0.75.0
+        pixi-version: v0.78.0
```
Comment (tvogels) [inline .github/workflows/skala-test.yml]: This version appears more than 10 times. Is there a way to reuse it as a constant?

File: skala/src/skala/pyscf/gradient_core.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
+# SPDX-License-Identifier: MIT
+
+"""Backend-independent PyTorch operations for PySCF nuclear gradients."""
```
Comment (jhrmnn) [inline skala/src/skala/pyscf/gradient_core.py:3]: It says "backend-independent" but is in `skala.pyscf`? Shouldn't there be some "neutral" ground that doesn't import either pyscf or gpu4pyscf by itself?

File: skala/src/skala/pyscf/gradient_core.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    return tuple(
+        torch.zeros_like(feature) if gradient is None else gradient.detach()
+        for feature, gradient in zip(differentiable_features, gradients, strict=True)
+    )
+
+
+def grid_derivative_block(
+    derivatives: FeatureMap, grid_start: int, grid_end: int
+) -> FeatureMap:
+    """Slice grid-resolved feature derivatives to one atom's grid block.
```
Comment (tvogels) [inline skala/src/skala/pyscf/gradient_core.py]:
```suggestion
    """Select a slice of the feature derivatives from `grid_start` to `grid_end`.
```

File: skala/src/skala/pyscf/gradient_core.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+class _Direction(IntEnum):
+    """Cartesian direction indices used by feature and potential tensors."""
+
+    X = 0
+    Y = 1
+    Z = 2
+
+
+_AO_GRADIENT = slice(_AOComponent.X, _AOComponent.XX)
+_GRADIENT_COMPONENTS = (_AOComponent.X, _AOComponent.Y, _AOComponent.Z)
```
Comment (jhrmnn) [inline skala/src/skala/pyscf/gradient_core.py]: Aren't these two the same thing, written differently? Could we drop the tuple, and do `ao[_AO_GRADIENT][d]` instead of `ao[_GRADIENT_COMPONENT[d]]`? Or if there's some performance difference, at least derive one from the other and name them consistently?

File: skala/src/skala/pyscf/gradient_core.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+            feature: derivative[..., grid_start:grid_end]
+            for feature, derivative in derivatives.items()
+            if feature in _POINT_LAST_FEATURES
+        }
+    )
+    return block
+
+
+def contract_veff_block(ao: torch.Tensor, derivatives: FeatureMap) -> torch.Tensor:
+    """Contract one atom's AO values with XC feature derivatives.
```
Comment (tvogels) [inline skala/src/skala/pyscf/gradient_core.py]: Can you describe what this does in math? Something like

Takes derivatives of the energy wrt the functional's input features, and using the ao matrix, turns them into derivatives of the energy wrt '...'. I tried to derivate what that is from the return shape, but didn't understand easily.

**Review by tvogels — DISMISSED** [review]

File: skala/src/skala/pyscf/gradient_core.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+
+# Grid metadata is laid out with one row per point: coordinates have shape
+# (npoints, 3), while weights have shape (npoints,).
+_POINT_FIRST_FEATURES = (Feature.GRID_COORDS, Feature.GRID_WEIGHTS)
+
+# Electronic features keep spin and, for GRAD, Cartesian components before the
+# grid dimension: DENSITY and KIN have shape (2, npoints), and GRAD has shape
+# (2, 3, npoints).
+_POINT_LAST_FEATURES = (Feature.DENSITY, Feature.GRAD, Feature.KIN)
```
Comment (jhrmnn) [inline skala/src/skala/pyscf/gradient_core.py]: Can't these two be merged and likewise the dict comprehensions below?

File: skala/src/skala/pyscf/gradient_core.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+    """
+    if not feature_tensors:
+        return ()
+
+    differentiable_features = tuple(
+        tensor.detach().requires_grad_(True) for tensor in feature_tensors
+    )
+    exc = exc_func(*differentiable_features)
+    if not exc.requires_grad:
```
Comment (jhrmnn) [inline skala/src/skala/pyscf/gradient_core.py:96]: This is not expected behavior, right? Should we instead raise?

File: skala/src/skala/pyscf/gradient_core.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+    gradients = torch.autograd.grad(
+        exc,
+        differentiable_features,
+        create_graph=False,
+        retain_graph=False,
+        allow_unused=True,
+    )
+    return tuple(
+        torch.zeros_like(feature) if gradient is None else gradient.detach()
```
Comment (jhrmnn) [inline skala/src/skala/pyscf/gradient_core.py]: `gradient is None` when the model doesn't use the feature? Wouldn't it be safer if a model comes with a contract about which features it uses, so there's no space for errors?

File: skala/src/skala/pyscf/gradient_core.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+        grid_start: Start of this block in the full molecular feature tensors.
+        grid_end: End of this block in the full molecular feature tensors.
+
+    Returns:
+        Grid-resolved derivatives restricted to ``[grid_start:grid_end]``.
+    """
+    block = {
+        feature: derivative[grid_start:grid_end]
+        for feature, derivative in derivatives.items()
+        if feature in _POINT_FIRST_FEATURES
```
Comment (jhrmnn) [inline skala/src/skala/pyscf/gradient_core.py]: Why the if? Are we intentionally silently dropping features not in the explicit list?

File: pixi.toml [hunk 10]
```diff
@@ -10,7 +10,7 @@ platforms = [
   { name = "linux-64-cuda12", platform = "linux-64", cuda = "12" },
   { name = "linux-64-cuda13", platform = "linux-64", cuda = "13" },
 ]
-requires-pixi = ">=0.75,<0.76"
+requires-pixi = ">=0.78,<0.79"
```
Comment (jhrmnn) [inline pixi.toml:13]: Pixi version is updated throughout in code, but not in documentation

**Review by jhrmnn — DISMISSED** [review]

File: skala/src/skala/pyscf/gradient_core.py (hunk 4 above)
Comment (JensWehner) [reply skala/src/skala/pyscf/gradient_core.py]: They are very similar, but not the same the upper one is `0:2` while the latter one is `0,1,2` giving the same result, but the first one creates a view, the second a copy under numpy indexing rules.

File: skala/src/skala/pyscf/gradient_core.py (hunk 2 above)
Comment (JensWehner) [reply skala/src/skala/pyscf/gradient_core.py:3]: I would argue, yes and no. Gpu4pyscf imports lots of stuff from pyscf, so common functionality can arguably live in pyscf. I would say it depends on how much you see gpu4pyscf as its own thing vs a plugin.

File: skala/src/skala/pyscf/gradient_core.py (hunk 6 above)
Comment (JensWehner) [reply skala/src/skala/pyscf/gradient_core.py]: If "these two" refers to _POINT_FIRST_FEATURES  and _POINT_LAST_FEATURES, we cannot merge them, as the dimension where the gridpoints sit differs, so we have to index them differently.

File: skala/src/skala/pyscf/gradient_core.py (hunk 7 above)
Comment (JensWehner) [reply skala/src/skala/pyscf/gradient_core.py:96]: Good question, let me think/investigate.

File: skala/src/skala/pyscf/gradient_core.py (hunk 8 above)
Comment (JensWehner) [reply skala/src/skala/pyscf/gradient_core.py]: gradient can also be None if the gradient is zero, so if you want to take higher derivatives on it, without this it would error out. If it is needed here, I have to check.

File: skala/src/skala/pyscf/gradient_core.py (hunk 9 above)
Comment (JensWehner) [reply skala/src/skala/pyscf/gradient_core.py]: no, because the indexing is different some quantities have points first and some last.

File: pixi.toml (hunk 10 above)
Comment (JensWehner) [reply pixi.toml:13]: good catch.

File: skala/src/skala/pyscf/gradient_core.py (hunk 7 above)
Comment (JensWehner) [reply skala/src/skala/pyscf/gradient_core.py:96]: we raise now. :)

File: skala/src/skala/pyscf/gradient_core.py (hunk 8 above)
Comment (JensWehner) [reply skala/src/skala/pyscf/gradient_core.py]: we raise now :)

File: skala/src/skala/pyscf/gradient_core.py (hunk 6 above)
Comment (JensWehner) [reply skala/src/skala/pyscf/gradient_core.py]: reworked maybe this is clearer.

File: skala/src/skala/pyscf/gradient_core.py (hunk 4 above)
Comment (JensWehner) [reply skala/src/skala/pyscf/gradient_core.py]: reworked not needed anymore.

File: skala/src/skala/pyscf/gradient_core.py (hunk 5 above)
Comment (JensWehner) [reply skala/src/skala/pyscf/gradient_core.py]: Changed docstring and naming does that help?

File: .github/workflows/skala-test.yml (hunk 1 above)
Comment (JensWehner) [reply .github/workflows/skala-test.yml]: made a custom action that unifies that, much nicer, thanks.

**Review by tvogels — APPROVED** [review]
